File: packages/core/src/artifacts.ts
import { generateBinary } from "./assemble.js";
import type { GeneratorConfig } from "./config.js";

export type GeneratedManifest = {
  name: string;
  programId: string;
  admin: string;
  arch: string;
  payloadSize: number;
  schemaHash: string;
  elfSha256: string;
};

export type DopplerArtifacts = {
  binary: Uint8Array;
  manifest: GeneratedManifest;
};

/**
 * Build Doppler program artifacts in memory from a generator config.
 *
 * Emits an ELF binary and produces a manifest with schema and ELF checksums.
 */
export async function createDopplerArtifacts(config: GeneratorConfig): Promise<DopplerArtifacts> {
  const binary = generateBinary({
    admin: config.admin,
    payloadSize: config.layout.payloadSize,
    arch: config.arch,
  });

  const manifest: GeneratedManifest = {
    name: config.name,
    programId: config.programId,
    admin: config.admin,
    arch: config.arch,
    payloadSize: config.layout.payloadSize,
    schemaHash: `sha256:${await sha256(JSON.stringify(config.layout.fields))}`,
    elfSha256: `sha256:${await sha256(binary)}`,
  };

  return { binary, manifest };
}

async function sha256(content: string | Uint8Array): Promise<string> {
  const bytes = typeof content === "string" ? new TextEncoder().encode(content) : content;
  const digestInput = new Uint8Array(bytes.byteLength);
  digestInput.set(bytes);
  const hash = await globalThis.crypto.subtle.digest("SHA-256", digestInput);
  return [...new Uint8Array(hash)].map((byte) => byte.toString(16).padStart(2, "0")).join("");
}

File: crates/cli/src/artifacts.rs
use std::fs;
use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};

use crate::assemble::generate_binary;
use crate::config::GeneratorConfig;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GeneratedManifest {
    pub name: String,
    pub program_id: String,
    pub admin: String,
    pub arch: String,
    pub payload_size: u32,
    pub schema_hash: String,
    pub elf_sha256: String,
}

#[derive(Debug, Clone)]
pub struct DopplerArtifacts {
    pub binary: Vec<u8>,
    pub manifest: GeneratedManifest,
}

pub fn create_doppler_artifacts(
    config: &GeneratorConfig,
) -> Result<DopplerArtifacts, crate::assemble::AssembleError> {
    let binary = generate_binary(&config.admin, config.layout.payload_size, config.arch)?;

    let schema_json =
        serde_json::to_string(&config.layout.fields).expect("layout fields serialize");
    let manifest = GeneratedManifest {
        name: config.name.clone(),
        program_id: config.program_id.clone(),
        admin: config.admin.clone(),
        arch: config.arch.as_str().to_string(),
        payload_size: config.layout.payload_size,
        schema_hash: format!("sha256:{}", sha256_hex(schema_json.as_bytes())),
        elf_sha256: format!("sha256:{}", sha256_hex(&binary)),
    };

    Ok(DopplerArtifacts { binary, manifest })
}

pub fn write_doppler_artifacts(
    config: &GeneratorConfig,
    binary_file: impl AsRef<Path>,
    manifest_file: Option<&Path>,
) -> Result<GeneratedManifest, ArtifactError> {
    let artifacts = create_doppler_artifacts(config)?;
    write_file_ensuring_dir(binary_file.as_ref(), &artifacts.binary)?;

    let manifest_path = manifest_file.map(PathBuf::from).unwrap_or_else(|| {
        binary_file
            .as_ref()
            .parent()
            .map(|dir| dir.join("manifest.json"))
            .unwrap_or_else(|| PathBuf::from("manifest.json"))
    });

    let manifest_json = serde_json::to_string_pretty(&artifacts.manifest)? + "\n";
    write_file_ensuring_dir(&manifest_path, manifest_json.as_bytes())?;

    Ok(artifacts.manifest)
}

#[derive(Debug, thiserror::Error)]
pub enum ArtifactError {
    #[error("{0}")]
    Assemble(#[from] crate::assemble::AssembleError),
    #[error("Failed to write artifact: {0}")]
    Io(#[from] std::io::Error),
    #[error("Failed to serialize manifest: {0}")]
    Json(#[from] serde_json::Error),
}

fn sha256_hex(bytes: &[u8]) -> String {
    let digest = Sha256::digest(bytes);
    digest.iter().map(|byte| format!("{byte:02x}")).collect()
}

fn write_file_ensuring_dir(path: &Path, content: &[u8]) -> Result<(), std::io::Error> {
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent)?;
    }
    fs::write(path, content)
}
